app/controllers/report_controller.rb

```diff
@@ -24,24 +26,47 @@ def add_record
                           ram_capacity_per_unit: data['ram_capacity_per_unit'],
                           disks: data['disk'] || [],
                           gpus: data['gpu'] || [],
-                          min: data['min'],
-                          half: data['half'],
-                          max: data['max'],
                           platform: data['platform'],
-                          cloud_provider: data['cloud_provider'],
                           instance_type: data['instance_type'],
                           location: data['location'],
                           tags: data['tags'] || []
                          )
-      device.save
+      if device.valid?
+        device.cloud_provider = Boavizta.provider(device.platform)
+        device.min = Boavizta.carbon_for_load(device, 0)
+        device.half = Boavizta.carbon_for_load(device, 50)
+        device.max = Boavizta.carbon_for_load(device, 100)
+        device.save
+      else
+        render json: "Error(s) with payload data: #{device.errors.full_messages.join(', ')}"
+        return
+      end
     end
 
     report = Report.new(device_id: device.uuid,
-                        current: data['current']
+                        current: Boavizta.carbon_for_load(device, data['current_load'])
                        )
-    report.save
 
-    render json: "Report saved successfully: Current carbon usage of #{data['current']}kgCO2eq saved for #{@current_user ? @current_user.username : 'anonymous user'}."
+    if report.valid?
+      report.save
+    else
+      render json: "Error(s) with payload data: #{report.errors.full_messages.join(', ')}"
+      return
+    end
+
+    render json: "Report saved successfully: Current carbon usage of #{report.current}kgCO2eq saved for #{@current_user ? @current_user.username : 'an anonymous user'}'s device '#{device.display_name}'"
+  end
+
+  def new_name
```

Comment:
That's a good point, I welcome input on making things more Rails-y since it's my first Rails app and I don't want things to be too unconventional. As you say, it's not a vital change for this PR specifically, but it's something to note for future, especially if we continue developing this post-ISC.